### PR TITLE
fix `t5gemma` tests

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

### CODE REVIEW

File: src/transformers/models/t5gemma/modeling_t5gemma.py [hunk 1]
```diff
@@ -1112,7 +1112,7 @@ def __init__(self, config: T5GemmaConfig):
         self.model = T5GemmaModel(config)
         self.vocab_size = config.decoder.vocab_size
         self.lm_head = T5GemmaLMHead(config.decoder.hidden_size, self.vocab_size)
-        self.loss_type = "ForMaskedLMLoss"
+        self.loss_type = "ForMaskedLM"
```
Comment (ydshieh) [inline src/transformers/models/t5gemma/modeling_t5gemma.py:1115]: simply wrong previously

File: src/transformers/models/t5gemma/modeling_t5gemma.py [hunk 2]
```diff
@@ -1168,7 +1168,7 @@ def forward(
             config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
             (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
         """
-        if self.training and self.config._attn_implementation != "eager":
+        if self.training and self.config._attn_implementation != "eager" and not is_torchdynamo_compiling():
```
Comment (ydshieh) [inline src/transformers/models/t5gemma/modeling_t5gemma.py]: @gante is this good approach?

Comment (vasqu) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: Imo shouldn't be too bad especially since it's just a warning here. I am a bit surprised that they suggest `eager`. How did you detect this one (just for fmi)?

File: tests/models/t5gemma/test_modeling_t5gemma.py [hunk 3]
```diff
@@ -596,6 +596,9 @@ class T5GemmaModelTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMi
     # used in `test_torch_compile_for_training`
     _torch_compile_train_cls = T5GemmaForConditionalGeneration if is_torch_available() else None
 
+    # won't fix
+    test_torchscript = False
```
Comment (vasqu) [inline tests/models/t5gemma/test_modeling_t5gemma.py:602]: Haha, we had that one more than once already :D

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/t5gemma/modeling_t5gemma.py:1115]: Surprised no test caught this (or did it), might need another look into adding/adjusting a test here

**Review by vasqu — APPROVED** [review]
LGTM! I'm surprised the loss one wasn't caught previously, should be checked at another time to make sure we have an easier time in the future

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: Since it's strongly recommended, maybe we want it to throw an exception when it is compiling? (= keep the original) 👀 

It is a good approach to skip warnings/exceptions, yes.

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/models/t5gemma/modeling_t5gemma.py:1115]: good point. not catching error because just a warning

```
        if loss_type is None or loss_type not in LOSS_MAPPING:
            logger.warning_once(
                f"`loss_type={loss_type}` was set in the config but it is unrecognised."
                f"Using the default loss: `ForCausalLMLoss`."
            )
```

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 2 above)
Comment (ydshieh) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: > How did you detect this one

just look the daily CI reports (mostly the new failure tests report :-))

Comment (ydshieh) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: > Since it's strongly recommended, maybe we want it to throw an exception when it is compiling? (= keep the original) 👀

faire enough. I will try

Comment (ydshieh) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: @gante with the last commit https://github.com/huggingface/transformers/pull/39052/commits/aac83967e998f4dfcb7c877f7692f57395652aa5

it could fail (as expected) but the message is super unclear and can't show the desired error message  provided by us.

Do you know why?

short version

```
E torch._dynamo.exc.Unsupported: Observed exception (EXCEPT_HANDLER) E Explanation: Dynamo found no exception handler at the top-level compiled function when encountering an exception. Exception will propagate outside the compiled region. This graph break is unexpected. E Hint: This is likely to be a Dynamo bug. Please report an issue to PyTorch. E E Developer debug context: raised exception ExceptionVariable(<class 'ValueError'>)
```

long version
```
>       loss = model(**inputs).loss

tests/test_modeling_common.py:4447:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py:1751: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py:1762: in _call_impl
    return forward_call(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ()
kwargs = {'attention_mask': tensor([[1, 1, 1, 1, 1, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0'), ... 82, 73,  3, 79, 92, 85]], device='cuda:0'), 'position_ids': tensor([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]], device='cuda:0')}
prior = None, cleanups = [<function nothing at 0x7f3a56c78700>], prior_skip_guard_eval_unsafe = False, saved_dynamic_layer_stack_depth = 0, cleanup = <function nothing at 0x7f3a56c78700>

    @functools.wraps(fn)
    def _fn(*args, **kwargs):
        prior = set_eval_frame(None)
        try:
            if is_fx_tracing():
                if config.error_on_nested_fx_trace:
                    raise RuntimeError(
                        "Detected that you are using FX to symbolically trace "
                        "a dynamo-optimized function. This is not supported at the moment."
                    )
                else:
                    return fn(*args, **kwargs)

            if is_jit_tracing():
                raise RuntimeError(
                    "Detected that you are using FX to torch.jit.trace "
                    "a dynamo-optimized function. This is not supported at the moment."
                )

            cleanups = [enter() for enter in self.enter_exit_hooks]
            prior_skip_guard_eval_unsafe = set_skip_guard_eval_unsafe(
                _is_skip_guard_eval_unsafe_stance()
            )

            # Ensure that if an assertion occurs after graph pushes
            # something onto the DynamicLayerStack then we pop it off (the
            # constructed graph code isn't guarded with try/finally).
            #
            # This used to be a context but putting a `with` here is a noticible
            # perf regression (#126293)
            saved_dynamic_layer_stack_depth = (
                torch._C._functorch.get_dynamic_layer_stack_depth()
            )
            _maybe_set_eval_frame(_callback_from_stance(callback))

            try:
                return fn(*args, **kwargs)
            except Unsupported as e:
                if config.verbose:
                    raise
>               raise e.with_traceback(None) from None
E               torch._dynamo.exc.Unsupported: Observed exception (EXCEPT_HANDLER)
E                 Explanation: Dynamo found no exception handler at the top-level compiled function when encountering an exception. Exception will propagate outside the compiled region. This graph break is unexpected.
E                 Hint: This is likely to be a Dynamo bug. Please report an issue to PyTorch.
E               
E                 Developer debug context: raised exception ExceptionVariable(<class 'ValueError'>)
E               
E               
E               from user code:
E                  File "/transformers/src/transformers/utils/generic.py", line 948, in wrapper
E                   if is_configured_to_return_tuple and is_top_level_module:
E               
E               Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"

/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py:659: Unsupported
--------------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------------It is strongly recommended to train T5Gemma models with the `eager` attention implementation instead of `sdpa`. Use `eager` with `AutoModelForCausalLM.from_pretrained('<path-to-checkpoint>', attn_implementation='eager')`.
============================================================================================================= short test summary info ==============================================================================================================FAILED tests/models/t5gemma/test_modeling_t5gemma.py::T5GemmaModelTest::test_torch_compile_for_training - torch._dynamo.exc.Unsupported: Observed exception (EXCEPT_HANDLER)
================================================================================================================ 1 failed in 4.44s =================================================================================================================root@fecdd565982b:/transformers# nano src/transformers/models/t5gemma/modeling_t5gemma.py
root@fecdd565982b:/transformers# HF_HOME=/mnt/cache RUN_SLOW=1 python3 -m pytest -v tests/models/t5gemma/test_modeling_t5gemma.py::T5GemmaModelTest::test_torch_compile_for_training
=============================================================================================================== test session starts ================================================================================================================
platform linux -- Python 3.10.12, pytest-8.4.1, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /transformers
configfile: pyproject.toml

```

Comment (vasqu) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: https://github.com/pytorch/pytorch/issues/154454 for reference, seems to be an internal issue where errors are not propagated as expected.

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/t5gemma/modeling_t5gemma.py:1115]: Could we add a strict test, ie some flag that would raise an error in those cases with an according test? Maybe not in this PR but I feel like it will avoid a lot of confusion down the line

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 2 above)
Comment (ydshieh) [reply src/transformers/models/t5gemma/modeling_t5gemma.py]: ok, thank you a lot. I will keep the value error here, although the trace is confusing, and adjust the test to be able to use `eager` (or other attn impl)

File: src/transformers/models/t5gemma/modeling_t5gemma.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/models/t5gemma/modeling_t5gemma.py:1115]: you mean a flag like `DEBUG` and if `DEBUG=True`, instead of ``warning_once, should raise error?

Comment (vasqu) [reply src/transformers/models/t5gemma/modeling_t5gemma.py:1115]: Yea something along those lines. If anyone knows that they should have a loss, it would be a nice way to catch it

File: tests/models/t5gemma/test_modeling_t5gemma.py [hunk 4]
```diff
@@ -595,6 +595,11 @@ class T5GemmaModelTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMi
 
     # used in `test_torch_compile_for_training`
     _torch_compile_train_cls = T5GemmaForConditionalGeneration if is_torch_available() else None
+    # `t5gemma` will give warning or raise error if it is not `eager` during training.
+    _torch_compile_train_attn_implementation = "eager"
```
Comment (vasqu) [inline tests/models/t5gemma/test_modeling_t5gemma.py:599]: This seems a bit too much to me :eyes: I think it makes more sense to overwrite the test here instead since it's such an edge case

Comment (vasqu) [reply tests/models/t5gemma/test_modeling_t5gemma.py:599]: More of a nit :D dont mind it

Comment (ydshieh) [reply tests/models/t5gemma/test_modeling_t5gemma.py:599]: yeah, i agree, but the change is also minimal, so i just go for it.

Comment (vasqu) [reply tests/models/t5gemma/test_modeling_t5gemma.py:599]: yup, it's fine no worries
